### DISCUSSION TITLE
Where can I find cert.pem and key.pem?

### DISCUSSION THREAD

**Discussion by TheTechRobo** [OpenShopChannel/WSC-Web#10] [Q&A] [answered]
I've been circumventing the cettificate addition but a few months ago (before it was added) I got an error in the WSC after changing the DNS and I suspect it's because HTTPS.

Thanks!

**Comment by spotlightishere** [comment]
Unfortunately a tricky question to answer. 😅

The EC component of the Wii Shop Channel uses the IOS SSL trust store, which by default is only Nintendo's CA. The main storefront as provided via Opera has its own trust store within the main ARC -> `opera/opcacrt6.dat`, only holding Nintendo's. Modifying Opera's trust store can be done via Opera on the desktop as well, which makes life a lot easier!

Our solution for IOS has been to hardcode our own certificate into the binary, patch WSC to load the cert to IOS, and go from there. However, this requires a specified length of a CA, and we've not even released our patches to the shop for things to function.

After some discussion with everyone, we think the best solution is to create a patcher to automate generating a certificate and inserting it, alongside separating other patches we've made along the way for others to review and modify. See [WSC-Patcher](https://github.com/OpenShopChannel/WSC-Patcher) for progress as we make this.

---
In the meantime - if you're interested in only the storefront, you'll need a 1024-bit CA with SHA-1 as the algorithm (SHA-256 will heavily break things, sadly.) I highly recommend utilizing Opera 9 on Windows to insert your own - you can copy and paste `opcacrt6.dat` from its place within %APPDATA%.
With Dolphin, you can disable SSL certificate verification to toy with EC as well.

Let me know if you have any questions, I'd love to help!

**Reply by TheTechRobo** [reply]
How do I generate a cert? Sorry - quite new to this kind of thing. ^^"

I've heard of using Certbot and Lets Encrypt, but I'm not qutie sure what that is.

Also - what do you mean the storefront? Do you mean just the HTML, i.e. no installation of WADs?

Finally - would Opera 9 for Linux work, and if so, can you give me a link for it? I can use WinXP but unfortunately it would take a while to get onto those PCs... they're also quite slow and frustrating to use. If you don,t have one, that's OK, I'll just try those PCs.

Thanks!

**Reply by spotlightishere** [reply]
No worries, hey!
After a little bit of testing, you should be able to with openssl:
```bash
# Generate CA certificate
openssl genrsa -out osc-ca.key 1024
openssl req -x509 -new -key osc-ca.key -days 3650 -nodes -subj "/CN=Open Shop Channel CA" -sha1 -out osc-ca.pem
```
Finally, create a certiifcate for your domain name:
```
openssl genrsa -out key.pem 1024
openssl req -new -key key.pem -out cert-to-sign.csr -nodes -subj "/CN=*.shop.your-domain.tld"
openssl x509 -req -in cert-to-sign.csr -CA osc-ca.pem -CAkey osc-ca.key -CAcreateserial -days 3650 -out cert.pem -sha1

rm cert-to-sign.csr
```
I haven't tested this, but this is roughly what we did to create originally. (You may also be interested in a tool like [XCA](https://hohnstaedt.de/xca/) to assist in creation!)

--- 
> Also - what do you mean the storefront? Do you mean just the HTML, i.e. no installation of WADs?

Ah yeah, sorry! The storefront is what WSC-Web provides, entirely HTML.

---
Regarding your last question - I _think_ it should; most likely any Opera version utilizing Presto should work, if you're able to find 12 or similar for Linux. I'm not quite sure where to download it, however.

I've used the Windows version via https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/ which helps out, it runs comfortably under Windows 10 thankfully!

**Reply by TheTechRobo** [reply]
TYSM, I'll try that soon!!

**Reply by TheTechRobo** [reply]
So I assume I use osc-ca.pem as the certificate?

Also, how would I add the new URL to the WAD?

Thanks again!!

**Reply by spotlightishere** [reply]
> So I assume I use osc-ca.pem as the certificate?

You got it, yupyup

> Also, how would I add the new URL to the WAD?

Oops - good point, forgot to mention that!
There're a few things you'll need to change:
 - Look for a string saying `Firmware %016llX required for %016llX can not be downloaded`. Shortly after that is `.shop.wii.com`. You'll want to change this to your domain, ensuring the bytes are exactly the same. This is strcmp'd to the end of your domain for some very simple check.
(If it'd help, `.a.taur.cloud` resolves to 127.0.0.1 for oss-auth, ecs, ccs, ucs, ias, [...])

 -  Similarly, for all instances of `https://oss-auth.shop.wii.com`, change them to `https://oss-auth.a.taur.cloud` or whatever your domain is!
 - Lastly, in the channel's main arc within the `opera` folder, you'll find `filter.ini`. You'll want to change `https://*.shop.wii.com` to your domain as well.

I _think_ that should be it, there may be more sneaking by.. you'll notice the channel displays a rather obscure 206xxx error (or less helpfully, a white screen) if it doesn't consider things secure which isn't very helpful :/

**Reply by TheTechRobo** [reply]
> you'll notice the channel displays a rather obscure 206xxx error

Yeah, _that's_ the one I got when I simply changed the DNS!

As a side note:

I don't really understand how certificates are secure. I can just make my own for any domain, and if I h ack into one, it would probably be trivial to replace the certificate. Edit: Oh, I just realised - it would prevent Man-In-The_Middle attacks.

But yeah, I'll try that later! (Though - how do I actually get _into_ the WAD? I tried searching for that string in a hex editor to no avail... Also, don't I have to keep the WAD the same size?)

**Reply by spotlightishere** [reply]
Pretty much regarding mitm - the server presents a certificate, and it's up to the client to accept or reject. Most devices have a CA list (say from [Mozilla's trust store](https://wiki.mozilla.org/CA)) where certificates presented can be verified against. Here Nintendo chose to only have their own, and we're overriding theirs with ours.

For the WAD - you can use a tool like [Sharpii](https://github.com/TheShadowEevee/Sharpii-NetCore) or [ShowMiiWads](https://wiibrew.org/wiki/ShowMiiWads) to extract its contents.

~~The main binary, most likely showing as `0000005e.app` or `00000001.app`, is LZ11 compressed - you can use [lzx](https://www.romhacking.net/utilities/826/) with `lzx -d <app name>`. You don't have to recompress it to repack in the WAD, and usually it makes editing a lot harder that way :P Decompressed, it's about 3.5 MB, if that helps to check if things are right.~~ This is wrong - it's not decompressed! I had another channel in mind when writing this, I'm sorry to confuse if it did.

The main ARC can be extracted using ShowMiiWads directly iirc or via [wszst](https://szs.wiimm.de/wszst/), whichever works best. Its name is `0000005f.app` and is about 3.4 MB, slightly smaller than the main binary.

You don't have to keep the WAD or the ARC the same size, but you do need to keep the binary the same size. Make sure that edits aren't longer than their original there - Nintendo uses `str*` for practically all of their string operations, so you can null pad your domains to fit if needed.

**Reply by TheTechRobo** [reply]
Wait - so I don't have to recompress it? Why does that make editing harder?

Also, wszst doesn't seem to allow compressing RARC, which I assume is an ARC?

**Reply by spotlightishere** [reply]
You don't have to decompress the dol specifically - it's a lot easier for quick changes as it's not required to compress. (Feel free to ignore if it helps out!)

I think wszst also calls calls an arc an `U8` archive. `wszst x 0000005f.app` should work to extract, and then `wszst c 0000005f.d` from the resulting folder to create once more.

**Reply by TheTechRobo** [reply]
????

Sorry for having so many problems. This is probably so annoying for you :/

![image](https://user-images.githubusercontent.com/52163910/136818142-616937b0-6f41-4e8e-b122-5651ae5e5b58.png)

**Reply by spotlightishere** [reply]
No worries at all! I should have been more specific - 0000005f.app is how it's installed to NAND (its content ID). Its 
index as Sharpii or similar might unpack it is `00000002.app`.

**Comment by spotlightishere** [comment, marked as answer]
Hey, I wanted to share that we now have a complete patcher utility available to automate provisioning a certificate, patching, and creating a usable WAD. If you get a chance, please give it a try and file issues!

https://github.com/OpenShopChannel/WSC-Patcher

**Reply by TheTechRobo** [reply]
Sure, thanks! I kept procrastinating your instructions - I hadn't even repacked the WAD yet, and  I haven't had my PC available for a while. Now I'm going to try that instead, thanks! :D

**Reply by TheTechRobo** [reply]
Hi, I just tried it. I copied `server.key` into WSC-Web's working dir as `key.pem` and `server.pem`, `cert.pem`. But when I `python3 app.py` I get `ssl.SSLError: [X509: KEY_VALUES_MISMATCH] key values mismatch (_ssl.c:4065)`. This is probably a stupid question, but how do I fix this? Thank you very much!

**Reply by spotlightishere** [reply]
Thank you for catching that, I was able to reproduce this locally - seems something changed during development, producing invalid keys 😬 I created [WSC-Patcher#2](https://github.com/OpenShopChannel/WSC-Patcher/issues/2) to track this.

In the meantime, you can create your own certificate via the `openssl` commands above, placing the root certificate in the `output` folder in DER form (i.e. `openssl x509 -pubout -outform DER -in osc-ca.pem`). See the [WSC-Patcher README](https://github.com/OpenShopChannel/WSC-Patcher#setup) for more info there!
